File: otfdvi.rb
#!/usr/bin/env ruby

require 'ostruct'
require 'rubygems'
require 'dvilib'
require 'optparse'
require 'yaml'
require 'fileutils'
require 'logger'
require 'minitest/autorun'

prgname = 'otfdvi'
prgfullname = 'Convert otfdvi (luaatex+fontspec) to dvips dvi'
prgversion = '0.0.1'
prgcredit = 'deimi@vtex.lt'

options = {}
optparse = OptionParser.new do|opts|
   # Set a banner, displayed at the top
   # of the help screen.
   opts.banner = "Usage: #{prgname} 
                  [OPTIONS] INPUT.DVI OUTPUT.DVI\n #{prgfullname}.\n  "
   opts.define_tail "\nEmail bug reports to #{prgcredit}"
   # Define the options, and what they do
   options[:debug] = false
   opts.on( '-d', '--debug', 'Print debug info [font tables, settings]' ) do
   options[:debug] = true
   end
   options[:verbose] = false
   opts.on( '--verbose', 'Be verbose' ) do
   options[:verbose] = true
   end
   options[:output] = "out.dvi"
   opts.on( '-c', '--output [FILE]', 'Write DVI output to [FILE]. 
                                      Default out.dvi' ) do |file|
   options[:output] = file
   end
   options[:dryrun] = false
   opts.on( '--dryrun', 'Do not perform actions.' ) do
   options[:dryrun] = true
   end
   options[:inplace] = false
   opts.on( '--inplaces', 'Output == Input' ) do
   options[:inplace] = true
   end
   options[:auto] = true
   opts.on( '--no-auto', 'Run make with all pdf' ) do
   options[:auto] = false
   end
   options[:htf] = true
   opts.on( '--no-htf', 'Do not generate htf fonts' ) do
   options[:htf] = false
   end
   options[:psmap] = "ttfonts.map"
   opts.on( '-u', '--psmap [FILE]', 'Output Postscript map file to [FILE]' ) do|file|
   options[:psmap] = file
   end

      opts.on( '-v', '--version', 'Print version' ) do
      puts "#{prgfullname}, v.#{prgversion}. #{prgcredit}"
      exit
   end
   options[:cfgfile] = prgname + '.yml'
   opts.on( '-c', '--cfgfile [FILE]', "read configuration file. 
                                       Default #{prgname}.yml" ) do|file|
   options[:cfgfile] = file
   end
   opts.on( '-h', '--help', 'Display this screen' ) do
     puts opts
     exit
   end
 end

begin
optparse.parse!
rescue OptionParser::InvalidOption => e
 puts e
 puts optparse
 exit 1
end

debug = options[:debug]
auto = options[:auto]
htf = options[:htf]


class Font
  attr_accessor :name
  attr_accessor :type # 0 - tfm; 1 - otf
  attr_accessor :otffilename, :script, :feature
  attr_accessor :charlist
  
  def initialize(name)
    @name = name
    @type = 0
    parsename(name)
    @charlist = Array.new()
  end

  def parsename(s)
    if s =~ /\[(.*.)\]:(.*?)/
      @type = 1
      @otffilename = $1
    end
    if s =~ /file:(.*?):script=(.*?);.*?;/
      @type = 1
      @otffilename = $1
      @script = $2
    end
  end
  
  def is_otf?()
    @type == 1 ? true : false  
  end

  def is_tfm?()
    @type == 0 ? true : false
  end

  def add_to_charlist(char)
    @charlist << char if not @charlist.include?(char)
  end
end

class TestFont < Minitest::Test
  def setup
    @font = Font.new("[lmroman10-regular]:+tlig;")
    @font2 = Font.new("file:lmroman17-regular:script=latn;+trep;+tlig;")
  end

  def test_if_otffont
    assert_equal true , @font.is_otf?
    assert_equal false, @font.is_tfm?
    assert_equal true, @font2.is_otf?
  end

end


TestFont.new("test1").run if debug

def otftocss(otf, font)
  f = Hash.new
  g = Hash.new
  run = "luaotfload-tool -n --find=file:#{otf} -i"
  result = Kernel.`(run)#`
  result.lines.each do |l|
    if l =~ /\s*(.*?):\s*(.*?)\s*\n/
      f[$1] = $2 if f[$1].nil?
    end
  end
 
  puts otf
  #puts f
  
  g["font-weight"] = f["weight"] if f["weight"] != "normal"
  g["font-style"] = "italic"  if f["italicangle"].to_i != 0
  g["font-family"] = "monospace" if f["monospaced"] == "true"
  puts g
  x = "htfcss: " + font + " " + g.map{|i,j| i + ": " + j + ";"}.join(" ")
  return x
end

logger = Logger.new(STDOUT)
logger.datetime_format = '%Y-%m-%d %H:%M:%S'

fileout = options[:output]
ymlfile = File.join(File.dirname(__FILE__), options[:cfgfile])
config = YAML.load_file(ymlfile)


mktempf = "otfdvi.Makefile"
mktemp = File.read(mktempf)

if !File.exist?(ymlfile)
 puts "Can't find config file #{ymlfile}. Aborting"
 exit
end
logger.info("Read config file: #{ymlfile}")

filein = ARGV[0]
fileout = ARGV[1] || fileout
fileout = filein if options[:inplace]
fileoutbase = File.basename(fileout, ".*")
psfileout = fileoutbase + ".ps"
pdffileout = fileoutbase + ".pdf"


logger.info("File in: #{filein}")
logger.info("File out: #{fileout}")
aglyphlist = Hash.new
adobeglyphlistfile = config[:adobe_glyph_list]
f = File.readlines(adobeglyphlistfile)
f.each do |l|
  next if l=~/^#/
  l=~/(.*?);(.*?)\n/
  aglyphlist[$2] = $1
end
logger.info("Read adobe glyph list: #{adobeglyphlistfile}")

dvi = Dvi.parse(File.open(filein, "rb"))
logger.info("Parse DVI file: #{filein}")




fonts = Hash.new
## Hash
## font = Font.new()
## font.id
## font.name
## font.type
## font.is_otf? .is_tfm?
## font.otffilename
## font.charlist == charset

currentfontnum = 0

dvi_modified = Array.new()
dvi.each do |op|
  #  puts op.inspect
  if op.class == Dvi::Opcode::Pre
    now = Time.now.strftime("%Y.%m.%d:%H%M")
    op.comment = " #{prgname} output #{now}"
  end

  if op.class == Dvi::Opcode::FntDef
    fontnum = op.num
    fonts[fontnum] = Font.new(op.fontname) if fonts[fontnum].nil?
    puts op.fontname
  
    if fonts[fontnum].is_otf?
      puts "#{op.fontname} => font#{fontnum}" if debug
      op.fontname = "font#{fontnum}" 
    end
  end
     
  if op.class == Dvi::Opcode::FntNum
    currentfontnum = op.index
  end

  if op.class == Dvi::Opcode::SetChar || op.class == Dvi::Opcode::Set
    if fonts[currentfontnum].is_otf?
      fonts[currentfontnum].add_to_charlist(op.index)
      i = fonts[currentfontnum].charlist.index(op.index)
      op.index = i
    end
  end

  dvi_modified << op
  
end


fx = File.open(fileout, "wb")
Dvi.write(fx, dvi_modified)
logger.info("Write DVI(modified) file: #{fileout}")

psmapfile = options[:psmap]

mkfile = config[:Makefile]
mk = File.open(mkfile, "w")

otffiles = Array.new
encfiles = Array.new
tfmfiles = Array.new
htffiles = Array.new
tfmtargets = Array.new

target_tmp=<<END
%{fontname}.tfm: %{otffontname} .FORCE
	otftotfm --literal-encoding=%{fontname}.enc --vendor=UKWN   $(verbose) --name=%{fontname} --map-file=%{psmapfile}  --no-updmap --warn-missing %{otffontname} 
END

fonts.keys.each do |fontid|
  if fonts[fontid].is_otf?
    puts "OTF: " + fonts[fontid].name if debug
    fontname  = "font#{fontid}"
    tfmfiles << fontname + ".tfm"
    otffontname = fonts[fontid].otffilename + ".otf"
    otffiles << otffontname
  
    encname = fontname
    encfile = encname + ".enc"
    encfiles << encfile
    htffile = fontname + ".htf" if htf
    htffiles << htffile if htf
    fh = File.open(encfile, "w")
    fhi = File.open(htffile, "w") if htf
    fh.puts "/#{encname} ["
    fhi.puts "#{fontname} 0 #{fonts[fontid].charlist.size-1}" if htf
    x = fonts[fontid].charlist.map{|i| "/" + aglyphlist["%04X" % i] }
    y = Array.new(256){|i| x[i] || "/.notdef"}
    fh.puts y
    fhi.puts fonts[fontid].charlist.map{|i| "'&#x" + "%04X" % i + ";' '' #{i.chr}(#{aglyphlist["%04X" % i]})" } if htf
    fh.puts "] def"
    fhi.puts "#{fontname} 0 #{fonts[fontid].charlist.size-1}" if htf
  
    cssstring = otftocss(otffontname, fontname) if htf
    fhi.puts cssstring if htf
    fh.close
    fhi.close if htf
  
    tfmtarget = target_tmp % { :fontname => fontname, :otffontname => otffontname , :psmapfile => psmapfile, :glyphlist => adobeglyphlistfile}
  
    tfmtargets << tfmtarget

    logger.info("Write encoding file: #{encfile} for #{otffontname}")
    logger.info("Write hyper text font file: #{htffile}") if htf
  else
    puts "TFM: " + fonts[fontid].name if debug
  end
end  

output = File.basename(fileout, ".*")
mkbody = mktemp % { :otffonts => otffiles.join(" "),
                    :encfiles => encfiles.join(" "),
                    :tfmfiles => tfmfiles.join(" "),
                    :htffiles => htffiles.join(" "),
                    :psmapfile => psmapfile,
                    :output => output,
                    :verbose => options[:verbose] ? '--verbose' : '' ,
}

mk.puts mkbody
mk.puts tfmtargets
  
mk.close
logger.info("Write Makefile: #{mkfile}")

if auto then
  run = "make -f #{mkfile} all pdf clean"
  logger.info("run system: #{run}")
  x = Kernel.exec(run)#`
  File.unlink mkfile
else
  run = "make -f #{mkfile} all "
  logger.info("run system: #{run}")
  x = Kernel.`(run)#`
end

File: otfdvi.lua
#!/usr/bin/env texlua


--[[
otfdvi program. Author Deimantas Galcius deimantas(dot)galcius(at)gmail(dot)com
--]]


otfdvi  = otfdvi or { }
local version = "0.1"
otfdvi.version = version
otfdvi.self = "otfdvi"

local _name = "otfdvi"
local exit = os.exit
local kpse = kpse
local kpse_version = kpse.version()
kpse.set_program_name("luatex")

local inspect  = require("inspect")
--print(inspect(package.loaded, {depth = 3}))
--print(inspect(file.dirname(lua.startupfile), {depth = 1}))

require("l-lpeg")
require("l-file")
local utils = require(file.dirname(lua.startupfile) .."/otfdvi/utils")
local flsparse        = utils.parse_fls
local reverse_table   = utils.reverse_table
local output_enc      = utils.output_enc
local write_glyphlist = utils.write_glyphlist
--
local dvi      = require("dvi")
local lustache = require("lustache")

local options = {
   filein = "",
   fileout = "out.dvi",
   psmapfile = "ttfonts.map",
   mkfile = "__Makefile",
   fontprefix = "font", 
   htf = false,
   debug = false,
   verbose = false,
   config = "otfdvi.conf.lua",
   --   outputdir = "./.otfdvi/",
   outputdir = "",
   dryrun = false,
   inplace = false,
   auto = false,
}

getopt = require("alt_getopt")

local function do_print_version()
   print(_name .. " version " .. version .. " licence GNU GPL v2.0")
   print("report problems to https://github.com/dgalcius/otfdvi/issues")
   exit()
end

local function do_print_help()
   print("Print help")
   exit()
end

local long_opts = {
   verbose = "V",
   help    = "h",
   auto    = "a",
   config  = "c",
   debug   = "d",
   fontprefix = "f",
   htf     = "H",
   log     = "l",
   mkfile  = "m",
   psmap   = "p",
--   len     = 1,
   output  = "o",
--   set_value = "S",
--   ["set-output"] = "o"
   version = "v"
}

local optarg
local optind
opts,optind,optarg = getopt.get_ordered_opts (arg, "hVvadHo:n:S:c:f:l:m:p:", long_opts)
      

for i, option in ipairs (opts) do

   if option == 'v' then
      do_print_version()
   end
   if option == 'h' then
      do_print_help()
   end

   if option == 'a' then
      options.auto = true
   end

   if option == 'd' then
      options.debug = true
   end

   if option == 'H' then
      options.htf = true
   end

   if option == 'c' then
      options.config = optarg[i]
   end

   if option == 'f' then
      options.fontprefix = optarg[i]
   end

   if option == 'l' then
      options.logfile = optarg[i]
   end

   if option == 'm' then
      options.mkfile = optarg[i]
   end

   if option == 'p' then
      options.psmapfile = optarg[i]
   end

   if option == 'o' then
      options.fileout = optarg[i]
   end

end

local j = 1
local arg_left = {}
for i = optind,#arg do
   arg_left[j] = arg[i]
   j = j + 1
end

options.filein = arg_left[1]
options.fileout = arg_left[2] or options.fileout
options.logfile = options.logfile or file.nameonly(options.filein) .. ".otfdvi.log"

local filein  = options.filein
local fileout = options.fileout
local logfile = options.logfile
local flsfile = file.nameonly(options.filein) .. ".fls"
local log = assert(io.open(logfile, 'w'))
local logw = function(...) log:write(...) end
logw("---", " !", "current time", "\n")
local outputdir = options.outputdir
local psmapfile = options.psmapfile
local mkfile = options.mkfile
local fontprefix = options.fontprefix
local htf = options.htf
local debug = options.debug
local verbose = options.verbose
local conf_file = options.config
logw("options:\n", inspect(options), "\n")

local dirname = file.dirname(arg[0])
local settings = require(file.join(dirname, conf_file))
logw("settings:\n", inspect(settings), "\n")

logw("kpse_version: ", kpse_version, "\n")
logw("lua_font_dir: ", lua_font_dir, "\n")

local fhi = assert(io.open(filein, 'rb'))
local fho = assert(io.open(fileout, 'wb'))
local psf = assert(io.open(psmapfile, 'w'))
local mkf = assert(io.open(mkfile, 'w'))

local fls = nil
if file.is_readable(flsfile) then
   fls = flsparse(flsfile)
end
local content = dvi.parse(fhi) -- get dvi file as lua table
local dvimodified = {}

function debug_print(s)
   if debug then
      print(s)
   end
end

local currfontnum = 0
local otffonts = {}
local ifonts = {}
local otf = false -- is_otf?
local fontdata = {}
local is_body = true --until opcode = post

logw("********** OPCODES ************** \n")
for _, op  in ipairs(content) do

   if op._opcode == "post" then
     is_body = false
   end
   
   if op._opcode == "pre" then
      op.comment = " otfdvi output " .. os.date("%Y.%m.%d:%H%M")
      debug_print(op.comment)
   end

   if op._opcode == "fntdef" then
      debug_print(inspect(op))
      logw(inspect(op), "\n")
      if is_body then
         otf, fontdata = getfontdata(op.fontname, op.design_size, op.scale, fls, { ["verbose"] = verbose } )
--         print(otf, inspect(fontdata))
--         os.exit(9)
         if otf then
            debug_print(op.fontname .. " => " .. fontprefix .. tostring(op.num))
            logw(op.fontname .. " => " .. fontprefix .. tostring(op.num), "\n")
            if otffonts[op.num] == nil then
               otffonts[op.num] = { fontname = op.fontname,
                                    basename = fontdata.filename,
                                    charlist = {},
                                    design_size = (op.design_size / 65536),
                                    design_size_dvi = op.design_size,
                                    scale = (op.scale / 65536),
                                    otfdata = fontdata,
               }
               ifonts[op.num] = { charlist = {} }
               ifonts[op.num].index = 0
            end
            op.fontname = fontprefix .. tostring(op.num)
         end
      else
         op.fontname = fontprefix .. tostring(op.num) -- # in postamble
      end
   end

   if op._opcode == "fntnum" or op._opcode == "fnt" then
      currfontnum = op.index
   end

   if op._opcode == "setchar" or op._opcode == "set" then
      if (otffonts[currfontnum] and true) then -- if otf font
         charlist = otffonts[currfontnum].charlist
         uclist = ifonts[currfontnum].charlist
         index = ifonts[currfontnum].index

         if uclist[op.index] == nil then -- char is not in a list yet
            table.insert(charlist, op.index)
            uclist[op.index] = index
            index = index + 1
            otffonts[currfontnum].charlist = charlist
            ifonts[currfontnum].charlist = uclist
            ifonts[currfontnum].index = index 
         end
         
         op.index = ifonts[currfontnum].charlist[op.index]
      end
   end

   table.insert(dvimodified, op)
end
logw(" ****** STOP OPCODES ****** \n")

debug_print(inspect(otffonts))
dvi.dump(fho, dvimodified) -- write modified DVI

logw("Output: ", fileout, "\n")
print("Written to " .. fileout)

--[[
- read '<otf-font-file>.lua'
--]]

for i_s, v_s in pairs(otffonts) do
   logw(" *** OTFFONT *** \n")   
   logw(inspect(v_s), "\n")
   local _resources, _runi, _metadata = {}, {}, {}
   local _f = nil
   local ghfilename = fontprefix .. i_s .. ".glyphs"
   _f = assert(dofile(v_s.otfdata.luafont))
   _resources = _f.resources
   _coverage = {}
   for _, _feat in ipairs(_f.resources.sequences) do
      if type(_feat.features) == "table" then
         if _feat.features["ssty"] or nil then
            _coverage = _feat.steps[1].coverage
         end
      end
   end

   local unicodes = {}
   for _i, _j in ipairs (v_s.charlist) do
      if _coverage[_j] then
         _j = _coverage[_j][1]
      end
      v_s.charlist[_i] = _j
   end
   for _i, _j in ipairs (v_s.charlist) do
      unicodes[_j] = "uni" .. string.format("%04X",_j)
   end

   _runi = reverse_table(_resources.unicodes)
   _metadata = _f.metadata

   for _i, _j in ipairs (otffonts[i_s].charlist) do
      unicodes[_j] = _runi[_j] or unicodes[_j]
   end
   logw("unicodes",  inspect(unicodes), "\n")

   write_glyphlist(ghfilename, reverse_table(unicodes))
   otffonts[i_s].glyphlist = reverse_table(unicodes)
   otffonts[i_s].metadata = metadata
   otffonts[i_s].glyphsfile = ghfilename
   otffonts[i_s].otffontnamefull = _resources.filename
   otffonts[i_s].otffontname = file.basename(_resources.filename)
   otffonts[i_s].otffonttype = file.suffixonly(_resources.filename)

   logw(" *** END OTFFONT *** \n")   
end

function XXoutput_htf(fontname, list, glyphlist, metadata)
   local file = fontname .. ".htf"
   local fh = assert(io.open(file, 'w'))
   local ss = fontname .. " 0 " .. (#list-1) .. '\n'
   local s = ss -- start
   for i, j in ipairs(list) do
      s = s .. "'&#x" .. string.format("%04X",j) ..";' ''  "
      s = s .. string.format("    %-12s",glyphlist[j]) .. " " .. " " .. string.format("%-5s\n",j)
   end
   s = s .. ss -- stop
   css = cssinfo(metadata)
   htfcss = "htfcss: " .. fontname .. " "
   for i, j in pairs(css) do
      htfcss = htfcss .. i .. ": " .. j .. "; "
   end
   s = s .. "\n" .. htfcss
--   print(htfcss)
--   os.exit()
   fh:write(s)
   io.close(fh)
   return true
end

view = {} -- view model for lustache
view.output = fileout
view.psmapfile = psmapfile
view.verbose = verbose
view.targets = {}
for i, v in pairs(otffonts) do
   otffonttype = v.otffonttype
   otffontname = v.otffontname
   fullpath = v.otffontnamefull
   glyphsfontname = v.glyphsfile
   design_size = v.design_size
   scale = v.scale
   script = v.otfdata.script
   mode = v.otfdata.mode
   language = v.otfdata.language
   feature = v.otfdata.features
   fontname = fontprefix .. i
   table.insert(view.targets,
                {       fontname = fontname,
                     otffontname = otffontname,
                        fullpath = fullpath,
                  glyphsfontname = glyphsfontname,
                  design_size = design_size,
                          scale = scale,
                          script = script,
                          mode = mode,
                          language = language,
                          feature = feature,
                          otffonttype = otffonttype,
                }
   )

   output_enc(fontname, v.charlist, v.glyphlist)
   htf = htf and output_htf(fontname, v.charlist, v.glyphlist, v.metadata)
end

-- template for lustache
tmpl = file.join(dirname, settings['mkfile_tmpl'])
local lm = assert(io.open(tmpl, 'r'))
local tmpl = lm.read(lm, '*all')

output = lustache:render(tmpl, view)
print("*****")
mkf:write(output)
mkf:close()
logw("Written to: ", mkfile, "\n")
print("Written to " .. mkfile, "\n")

